### ISSUE TITLE
SoftMax giving particular trouble in book example.

### ISSUE THREAD

**Issue by bdleitner** [gorgonia/gorgonia#447]
TL:DR;
When trying to run the example code from [Hands On Deep Learning with Go, chapter 3](https://github.com/PacktPublishing/Hands-On-Deep-Learning-with-Go/blob/master/Chapter03/mnist_feedforward/main.go), learnables end up at infinity.

Full version:
First off, apologies if this isn't really an issue.  It's entirely possible that I'm just using things wrong or that the example in the book is bad (one would have hoped that it had been tested, but maybe things have changed in the interim?  
In any case, some of the suspicious behavior I observed was not linked to SoftMax, so I'm less sure that it's the only source of weirdness here.  Nevertheless, something is working which is somewhat unexpected from a straight-out-of-the-book example, and I wasn't sure where else to look for help specific to this library and its usage.  So if this is a poor choice, my apologies again and if you point me in the right direction I'd greatly appreciate it. 

Anyway, after having dabbled some in ML last year (and filing issue #323), I decided to try and actually learn more of this from scratch.  I got both "Hands On Deep Learning with Go" (HODLwG) and "Go Machine Learning Projects" to try and do some learning.

The first nontrivial example in HODLwG is the typical MNIST handwriting example.  I am able to reproduce this problem with code directly copied from the example, but I'm including `main.go` file, in which I've taken out some of their file-writing, progress bars, and other cruft and added some more options for logging:

Here's what I've found.  When using the Hadamard Product cost function as in the example, within ~10 batches or so, predictions go to +Inf (See attached [example1.txt](https://github.com/gorgonia/gorgonia/files/5435755/example1.txt))
This seems to happen as soon as the one-hot output vectors actually get close enough to having a singular 1 value.
It is also quite suspicious that by batch 5 every prediction is the same (all predictions in batch 5 are for 9s, batch 6 is all 1s, then we get all 2s, all 0s, all 4s, all 7s, and finally, while batch 11 has Infs, all non-Inf predictions are for 9s).  I wondered if the learning rate (default of 0.001) was somehow too high and it was deciding that "there were more Xs in batch n than anything else so maybe it's a good strategy to always guess x), but I didn't actually see this validated.  Indeed, while all non-Inf predictions in batch 11 were 9s, there weren't actually any true 9s in batch 10.

Trying to use `StableSoftMax` to see if that would help but that just resulted in (reminders of #323, though it seems to be a different error):
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x88 pc=0xdbb321]

goroutine 1 [running]:
gorgonia.org/gorgonia.Read(0x0, 0xc000092268, 0xff6ba0)
	C:/projects/go/bdleitner/src/gorgonia.org/gorgonia/gorgonia.go:300 +0x161
main.(*nn).fwd(0xc000092240, 0xc000142000, 0xe83ac0, 0xc00324beb8)
	C:/projects/go/bdleitner/src/github.com/bdleitner/dlwg/chapter3/feedforward/main.go:93 +0x1e5
main.main()
	C:/projects/go/bdleitner/src/github.com/bdleitner/dlwg/chapter3/feedforward/main.go:124 +0x40c
```

I tried manually subtracting the `Max` before taking `SoftMax`.  This prevented the Inf problem, at least in small doses.  I tried runnign two epochs at batch size 100.  The end result was very poor performance and then I looked at the output logs, the vast majority of predictions were 2s, and everything that was not predicted to be a 2 was predicated to be a 9.  I thought maybe it just needed more time, so I tried running for 20 epochs, but it crapped out at Epoch 6, batch 227 when cost became -Inf.

Trying manually subtracting `Mean(out)` instead of `Max(out)` gave the curious result that predictions all converged toward flat vectors of 0.1s.

I then tried changing from SoftMax to Sigmoid.  This wouldn't get me one-hot vectors but I wanted values between 0 and 1 so it would at least do that.  To my surprise I still ended up with predictions all converging to a single value (in my case 4).  Unlike with SoftMax, the predictions did not jump around and stayed consistently on 4 one that was reached.

Since that obviously had nothing to do with SoftMax I tried switching the cost function to `BinaryXEnt` instead.  This actually worked and gave me a network that was able to learn (after 5 epochs, I had ~80% success rate in testing).

So I tried going back to SoftMax but using BinaryXEnt instead.  This seems dangerous as when you actually get to singular values in the one-hot the logs on the 0s explode... and with actual_y in this case being 0.1 on the desired "off" values, I expected this to go crazy.  I did in fact end up with Infs again and the cost was NaN.  Leading up to that the predictions had converged that everything was a 0.  In this case, however, the confidence was lower.  The max value in the outputs was always in the 0 index, but was generally in the ~0.25-0.35 range when we hit NaN cost and crapped out.

Here's my `main.go`:
```
package main

import (
  "flag"
  "fmt"
  "log"
  "math"
  "math/rand"
  "strconv"
  "strings"

  "github.com/pkg/errors"
  "gorgonia.org/gorgonia"
  "gorgonia.org/gorgonia/examples/mnist"
  "gorgonia.org/tensor"
)

var (
  epochs       = flag.Int("epochs", 100, "Number of epochs to train for")
  dataset      = flag.String("dataset", "train", "Which dataset to train on? Valid options are \"train\" or \"test\"")
  batchsize    = flag.Int("batch_size", 100, "Batch size")
  cpuprofile   = flag.String("cpuprofile", "", "CPU profiling")
  batchLogging = flag.String("batch_logging", "none", "How much logging to info to output during each training batch.  Acceptable values are \"all\" and \"one\", anything else is treated as no logging")
  useXEnt      = flag.Bool("use_cross_entropy", false, "If true, uses BinaryXEnt for the losses function, otherwise uses simple Hadamard product")
)

// This will need adjusting to your own directory layout and the location of the MNIST files
// Note: I had to rename some of the files as the filenames were not quite as expected
// (some '.'s needed replacement with '-'s)
const loc = "../mnist/"

type nn struct {
  g          *gorgonia.ExprGraph
  w0, w1, w2 *gorgonia.Node

  out     *gorgonia.Node
  predVal gorgonia.Value
}

type sli struct {
  start, end int
}

func (s sli) Start() int { return s.start }
func (s sli) End() int   { return s.end }
func (s sli) Step() int  { return 1 }

func newNN(g *gorgonia.ExprGraph) *nn {
  // Create node for w/weight
  w0 := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithShape(784, 300), gorgonia.WithName("w0"), gorgonia.WithInit(gorgonia.GlorotN(1.0)))
  w1 := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithShape(300, 100), gorgonia.WithName("w1"), gorgonia.WithInit(gorgonia.GlorotN(1.0)))
  w2 := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithShape(100, 10), gorgonia.WithName("w2"), gorgonia.WithInit(gorgonia.GlorotN(1.0)))

  return &nn{
    g:  g,
    w0: w0,
    w1: w1,
    w2: w2,
  }
}

func (m *nn) learnables() gorgonia.Nodes {
  return gorgonia.Nodes{m.w0, m.w1, m.w2}
}

func (m *nn) fwd(x *gorgonia.Node) (err error) {
  var l0, l1, l2 *gorgonia.Node
  var l0dot, l1dot *gorgonia.Node

  // Set first layer to be copy of input
  l0 = x

  // Dot product of l0 and w0, use as input for ReLU
  if l0dot, err = gorgonia.Mul(l0, m.w0); err != nil {
    return errors.Wrap(err, "Unable to multiply l0 and w0")
  }

  // Build hidden layer out of result
  l1 = gorgonia.Must(gorgonia.Rectify(l0dot))

  if l1dot, err = gorgonia.Mul(l1, m.w1); err != nil {
    return errors.Wrap(err, "Unable to multiply l1 and w1")
  }
  l2 = gorgonia.Must(gorgonia.Rectify(l1dot))

  var out *gorgonia.Node
  if out, err = gorgonia.Mul(l2, m.w2); err != nil {
    return errors.Wrapf(err, "Unable to multiply l2 and w2")
  }

  // Optional tweaks to modify the use of SoftMax
  //out = gorgonia.Must(gorgonia.Sub(out, gorgonia.Must(gorgonia.Mean(out)))) // can also use Max here instead of Mean.
  m.out, err = gorgonia.SoftMax(out)
  //m.out, err = gorgonia.Sigmoid(out) // Sigmoid seems to work iff it's used with BinaryXEnt rather than Hadamard
  gorgonia.Read(m.out, &m.predVal)
  return

}

func main() {
  flag.Parse()
  rand.Seed(7945)

  // // intercept Ctrl+C
  // sigChan := make(chan os.Signal, 1)
  // signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
  // doneChan := make(chan bool, 1)

  var inputs, targets tensor.Tensor
  var err error

  // load our data set
  trainOn := *dataset
  if inputs, targets, err = mnist.Load(trainOn, loc, tensor.Float64); err != nil {
    log.Fatal(err)
  }

  numExamples := inputs.Shape()[0]
  bs := *batchsize

  g := gorgonia.NewGraph()
  x := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithShape(bs, 784), gorgonia.WithName("x"))
  y := gorgonia.NewMatrix(g, tensor.Float64, gorgonia.WithShape(bs, 10), gorgonia.WithName("y"))

  m := newNN(g)
  if err = m.fwd(x); err != nil {
    log.Fatal(err)
  }

  var cost *gorgonia.Node
  if *useXEnt {
    log.Println("Using Cross-Entropy")
    losses, err := gorgonia.BinaryXent(m.out, y)
    if err != nil {
      log.Fatal(err)
    }
    cost = gorgonia.Must(gorgonia.Mean(losses))
  } else {
    log.Println("Using Hadamard Product")
    losses, err := gorgonia.HadamardProd(m.out, y)
    if err != nil {
      log.Fatal(err)
    }
    cost = gorgonia.Must(gorgonia.Mean(losses))
    cost = gorgonia.Must(gorgonia.Neg(cost))
  }

  var costVal gorgonia.Value
  gorgonia.Read(cost, &costVal)

  if _, err = gorgonia.Grad(cost, m.learnables()...); err != nil {
    log.Fatal(err)
  }

  vm := gorgonia.NewTapeMachine(g, gorgonia.BindDualValues(m.learnables()...))
  solver := gorgonia.NewRMSPropSolver(gorgonia.WithLearnRate(0.00001), gorgonia.WithBatchSize(float64(bs)))

  batches := numExamples / bs
  log.Printf("Batches %d", batches)

  for i := 0; i < *epochs; i++ {
    for b := 0; b < batches; b++ {
      start := b * bs
      end := start + bs
      if start >= numExamples {
        break
      }
      if end > numExamples {
        end = numExamples
      }
      var xVal, yVal tensor.Tensor
      if xVal, err = inputs.Slice(sli{start, end}); err != nil {
        log.Fatal("Unable to slice x")
      }
      if yVal, err = targets.Slice(sli{start, end}); err != nil {
        log.Fatal("Unable to slice y")
      }
      if err = xVal.(*tensor.Dense).Reshape(bs, 784); err != nil {
        log.Fatalf("Unable to reshape %v", err)
      }

      gorgonia.Let(x, xVal)
      gorgonia.Let(y, yVal)
      if err = vm.RunAll(); err != nil {
        log.Fatalf("Failed at Epoch %d batch %d: %v", i, b, err)
      }

      arrayOutput := m.predVal.Data().([]float64)
      labels := yVal.Data().([]float64)
      fmt.Printf("Epoch %d, batch %d\n", i, b)
      logBatchInfo(arrayOutput, labels)
      cost := costVal.Data().(float64)
      fmt.Printf("Cost = %0.4f\n", cost)
      if math.IsNaN(cost) || math.IsInf(cost, 0) {
        fmt.Println("Cost is NaN or Inf, quitting.")
        return
      }
      fmt.Println()

      solver.Step(gorgonia.NodesToValueGrads(m.learnables()))
      vm.Reset()
    }
    log.Printf("Epoch %d | cost %v", i, costVal)
  }

  log.Printf("Run Tests")

  // load our test set
  if inputs, targets, err = mnist.Load("test", loc, tensor.Float64); err != nil {
    log.Fatal(err)
  }

  numExamples = inputs.Shape()[0]
  bs = *batchsize
  batches = numExamples / bs

  for b := 0; b < batches; b++ {
    start := b * bs
    end := start + bs
    if start >= numExamples {
      break
    }
    if end > numExamples {
      end = numExamples
    }

    var xVal, yVal tensor.Tensor
    if xVal, err = inputs.Slice(sli{start, end}); err != nil {
      log.Fatal("Unable to slice x")
    }

    if yVal, err = targets.Slice(sli{start, end}); err != nil {
      log.Fatal("Unable to slice y")
    }

    if err = xVal.(*tensor.Dense).Reshape(bs, 784); err != nil {
      log.Fatalf("Unable to reshape %v", err)
    }

    gorgonia.Let(x, xVal)
    gorgonia.Let(y, yVal)
    if err = vm.RunAll(); err != nil {
      log.Fatalf("Failed at epoch test: %v", err)
    }

    arrayOutput := m.predVal.Data().([]float64)
    labels := yVal.Data().([]float64)
    fmt.Printf("Testing batch %d\n", b)
    logBatchInfo(arrayOutput, labels)

    vm.Reset()
  }
  log.Printf("Epoch Test | cost %v", costVal)
}

func oneHot(values []float64) (int, float64) {
  max := math.Inf(-1)
  maxI := -1

  for i, v := range values {
    if v > max {
      maxI, max = i, v
    }
  }
  return maxI, max
}

func oneHotStats(size int, gots []float64, wants []float64) (int, int) {
  correct := 0
  total := 0
  for start := 0; start < len(gots); start += size {
    total++
    got, _ := oneHot(gots[start : start+size])
    want, _ := oneHot(wants[start : start+size])
    if got == want {
      correct++
    }
  }
  return correct, total
}

func oneHotString(values []float64) string {
  var outs []string
  for _, value := range values {
    outs = append(outs, strconv.FormatFloat(value, 'g', 5, 64))
  }
  index, max := oneHot(values)
  return fmt.Sprintf("[%s], %d, %g", strings.Join(outs, ", "), index, max)
}

func logBatchInfo(got, want []float64, opts ...string) { // opts is a quick and dirty hack to override the flag
  correct, total := oneHotStats(10, got, want)
  fmt.Printf("%d of %d correct\n", correct, total)
  var max int

  flag := *batchLogging
  if len(opts) > 0 {
    flag = opts[0]
  }

  switch strings.ToLower(flag) {
  case "all":
    max = len(got) / 10
  case "one":
    max = 1
  default:
    // no logging, just return
    return
  }

  for i := 0; i < max; i++ {
    fmt.Printf("Output got[%d] = %s\n", i, oneHotString(got[10*i:10*(i+1)]))
    fmt.Printf("Output want[%d] = %s\n", i, oneHotString(want[10*i:10*(i+1)]))
  }
}
```

**Comment by chewxy** [comment]
This and #446 are related to the upgrades we made to the softmax op. It should be fixed today
